### PR TITLE
[Integrations][Custom] bug fixes

### PR DESCRIPTION
**Pull request by matarpeles**
### **User description**
- Fix SSL verification: Respect verify_ssl config by creating custom httpx client when disabled
- Fix base URL path trimming: Preserve full base URL path when combining with endpoint paths
- Fix list response handling: Auto-detect direct list responses and set data_path to '.' if missing
- Add custom headers support: Global custom headers with env var resolution, per-endpoint headers override global
- Add comprehensive tests: Test cases for SSL verification, URL construction, custom headers, and list response handling

# Description

What -

Why -

How -

## Type of change

Please leave one option from the following and delete the rest:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] New Integration (non-breaking change which adds a new integration)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Non-breaking change (fix of existing functionality that will not change current behavior)
- [ ] Documentation (added/updated documentation)

<h4> All tests should be run against the port production environment(using a testing org). </h4>

### Core testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Resync finishes successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Scheduled resync able to abort existing resync and start a new one
- [ ] Tested with at least 2 integrations from scratch
- [ ] Tested with Kafka and Polling event listeners
- [ ] Tested deletion of entities that don't pass the selector

### Integration testing checklist

- [ ] Integration able to create all default resources from scratch
- [ ] Completed a full resync from a freshly installed integration and it completed successfully
- [ ] Resync able to create entities
- [ ] Resync able to update entities
- [ ] Resync able to detect and delete entities
- [ ] Resync finishes successfully
- [ ] If new resource kind is added or updated in the integration, add example raw data, mapping and expected result to the `examples` folder in the integration directory.
- [ ] If resource kind is updated, run the integration with the example data and check if the expected result is achieved
- [ ] If new resource kind is added or updated, validate that live-events for that resource are working as expected
- [ ] Docs PR link [here](#)

### Preflight checklist

- [ ] Handled rate limiting
- [ ] Handled pagination
- [ ] Implemented the code in async
- [ ] Support Multi account

## Screenshots

Include screenshots from your environment showing how the resources of the integration will look.

## API Documentation

Provide links to the API documentation used for this integration.

___

### **PR Type**
Bug fix, Enhancement, Tests

___

### **Description**
- Fix SSL verification: Create custom httpx client when verify_ssl is disabled

- Fix URL construction: Preserve full base URL path when combining with endpoints

- Fix list response handling: Auto-detect direct list responses and set data_path to '.'

- Add custom headers support: Global headers with env var resolution, per-endpoint override

- Add comprehensive test coverage: URL construction, custom headers, and list response handling

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["Client Initialization"] -->|verify_ssl disabled| B["Create Custom httpx Client"]
  A -->|custom_headers provided| C["Parse & Resolve Env Vars"]
  D["URL Construction"] -->|preserve path| E["Manual URL Concatenation"]
  F["Response Processing"] -->|detect list| G["Auto-set data_path to '.'"]
  F -->|merge headers| H["Global + Endpoint Headers"]
  B --> I["HTTP Client"]
  C --> I
  E --> I
  G --> I
  H --> I
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Bug fix</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>client.py</strong><dd><code>SSL verification, URL construction, and custom headers fixes</code></dd></summary>
<hr>

integrations/custom/http_server/client.py

<ul><li>Added <code>custom_headers</code> parameter to constructor with merging logic in <br><code>_make_request</code><br> <li> Replaced <code>urljoin</code> with manual URL construction to preserve full base <br>URL paths<br> <li> Added SSL verification handling: creates custom httpx client when <br><code>verify_ssl=False</code><br> <li> Removed generic exception handler in <code>_make_request</code> for cleaner error <br>handling<br> <li> Added validation for empty base_url and endpoint parameters</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2427/files#diff-2b82e0d6b02d32f4e41a6b3fff5bf55aae7911b7da426a8e675d1a54e3820766">+29/-10</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Auto-detect list responses and improve data_path handling</code></dd></summary>
<hr>

integrations/custom/main.py

<ul><li>Added auto-detection logic for direct list responses when <code>data_path</code> is <br>not specified<br> <li> Sets <code>data_path</code> to <code>'.'</code> when response is detected as a list<br> <li> Added error logging when response is not a list and <code>data_path</code> is <br>missing<br> <li> Moved <code>data_path</code> extraction inside endpoint loop for per-endpoint <br>handling<br> <li> Improved logging to guide users on specifying <code>data_path</code> when needed</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2427/files#diff-0fcca960cb3550c61820c8ad38abacd52cad18878ca87b24f32039eb331c0a03">+23/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>initialize_client.py</strong><dd><code>Custom headers parsing with environment variable resolution</code></dd></summary>
<hr>

integrations/custom/initialize_client.py

<ul><li>Added <code>_resolve_env_vars</code> function to resolve environment variable <br>references in format <code>${VAR_NAME}</code><br> <li> Added <code>_parse_custom_headers</code> function to parse JSON custom headers and <br>resolve env vars<br> <li> Integrated custom headers parsing into <code>init_client</code> factory function<br> <li> Uses Ocean's Pydantic <code>parse_raw_as</code> utility for consistent config <br>parsing</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2427/files#diff-511fea9e90c6ad3de90c59bf89f6c47ce21158b3cb915d212f5d9e2b1a58bd47">+41/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_client.py</strong><dd><code>Comprehensive URL construction and custom headers tests</code>&nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/tests/test_client.py

<ul><li>Added 11 comprehensive URL construction test cases covering various <br>scenarios<br> <li> Tests include trailing slashes, path components, port numbers, deep <br>paths, and edge cases<br> <li> Added 3 custom headers test cases verifying global and per-endpoint <br>header merging<br> <li> Tests validate header override behavior and default empty dict <br>initialization<br> <li> All tests use mocking to isolate client behavior from actual HTTP <br>requests</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2427/files#diff-f2344e94ef1e598d1e915f5cd1c98058bb96f0913dfd84177011ad675f6c680f">+313/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>test_main.py</strong><dd><code>List response handling and auto-detection tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

integrations/custom/tests/test_main.py

<ul><li>Added 3 test cases for list response auto-detection functionality<br> <li> Tests verify auto-detection of direct list responses and <code>data_path</code> <br>setting<br> <li> Tests validate error logging when response is not a list and <code>data_path</code> <br>is missing<br> <li> Tests confirm explicit <code>data_path</code> is used when provided (no <br>auto-detection)<br> <li> Uses mocking for client, resource config, and entity processor <br>isolation</ul>

</details>

  </td>
  <td><a href="https://github.com/port-labs/ocean/pull/2427/files#diff-4847816b477a95168229c6c6bdede60436edb6cd8855946418a773d2a264589b">+164/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

</details>

___

### CODE REVIEW

File: integrations/custom/http_server/client.py [hunk 1]
```diff
@@ -33,21 +32,33 @@ def __init__(
...
         # Use Ocean's built-in HTTP client with retry and rate limiting
-        self.client = http_async_client
+        # If SSL verification is disabled, create a custom client
+        if not verify_ssl:
+            logger.warning(
+                "SSL verification is disabled - not recommended for production"
+            )
+            self.client = httpx.AsyncClient(verify=False)
+        else:
+            self.client = http_async_client
```
Comment (omby8888) [inline integrations/custom/http_server/client.py]: This else is redundant

File: .DS_Store
Comment (omby8888) [inline .DS_Store:1]: pls delete those

File: integrations/custom/.port/resources/blueprints.json
Comment (omby8888) [inline integrations/custom/.port/resources/blueprints.json:1]: what's that blueprint?

File: integrations/custom/http_server/client.py [hunk 2]
```diff
@@ -33,21 +32,33 @@ def __init__(
...
+        self.custom_headers = custom_headers or {}
 
         # Use Ocean's built-in HTTP client with retry and rate limiting
-        self.client = http_async_client
+        # If SSL verification is disabled, create a custom client
+        if not verify_ssl:
+            logger.warning(
+                "SSL verification is disabled - not recommended for production"
+            )
+            self.client = httpx.AsyncClient(verify=False)
```
Comment (omby8888) [inline integrations/custom/http_server/client.py]:
```suggestion
            self.client = OceanAsyncClient(
                RetryTransport,
                timeout=ocean.config.client_timeout,
                verify=False
            )
```
that way u don't lose the retries and timeout configs

File: integrations/custom/http_server/client.py [hunk 3]
```diff
@@ -79,7 +90,19 @@ async def _fetch_with_pagination(
     ) -> AsyncGenerator[List[Dict[str, Any]], None]:
         """Fetch data with pagination handling using handler pattern"""
 
-        url = urljoin(self.base_url, endpoint.lstrip("/"))
+        # Fix: Preserve full base URL path when combining with endpoint
+        # urljoin removes last path component if endpoint starts with /
+        # So we manually construct the URL to preserve the full path
```
Comment (omby8888) [inline integrations/custom/http_server/client.py]: no need for the comment

File: integrations/custom/http_server/client.py [hunk 4]
```diff
@@ -79,7 +90,19 @@ async def _fetch_with_pagination(
...
+        base_url = self.base_url.rstrip("/")
+        endpoint_path = endpoint.lstrip("/")
+
+        # Validate - raise error if empty
```
Comment (omby8888) [inline integrations/custom/http_server/client.py:96]: same

File: integrations/custom/http_server/client.py [hunk 5]
```diff
@@ -121,12 +144,16 @@ async def _make_request(
         headers: Dict[str, str],
     ) -> httpx.Response:
         """Make HTTP request using Ocean's built-in client with retry and rate limiting"""
+        # Merge global custom headers with per-endpoint headers
+        # Per-endpoint headers override global headers (same key)
```
Comment (omby8888) [inline integrations/custom/http_server/client.py]: same

File: integrations/setup-wizard/frontend/.next/trace
Comment (omby8888) [inline integrations/setup-wizard/frontend/.next/trace:1]: remove

File: port_ocean/.DS_Store
Comment (omby8888) [inline port_ocean/.DS_Store:1]: remove

File: integrations/custom/test-config.yml
Comment (omby8888) [inline integrations/custom/test-config.yml:1]: why do we need those?

File: integrations/custom/test-config-local.yml
Comment (omby8888) [inline integrations/custom/test-config-local.yml:1]: why do we need those?

File: integrations/custom/main.py [hunk 6]
```diff
@@ -51,6 +53,25 @@ async def resync_resources(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
+                    else:
+                        # Not a list and no data_path specified - log error and continue
+                        # Let Ocean's validation handle it downstream
+                        logger.error(
+                            f"Response from {endpoint} is not a list and 'data_path' is not specified. "
+                            f"Yielding response as-is. If mapping fails, please specify 'data_path' in your selector "
+                            f"(e.g., data_path: '.data'). Response type: {type(response_data).__name__}"
+                        )
+                        # Continue without setting data_path - will yield batch as-is
+
```
Comment (omby8888) [inline integrations/custom/main.py]: can u explain why this section is needed instead of simply set `data_path = getattr(selector, "data_path", ".")`, and just before iterating the batch to check if it's a list otherwise yield as it is?

File: integrations/custom/.port/resources/blueprints.json
Comment (Matarp13) [reply integrations/custom/.port/resources/blueprints.json:1]: Fixed

File: integrations/custom/http_server/client.py (hunk 2 above)
Comment (Matarp13) [reply integrations/custom/http_server/client.py]: Fixed

File: integrations/custom/main.py (hunk 6 above)
Comment (Matarp13) [reply integrations/custom/main.py]: Fixed

File: integrations/custom/http_server/client.py [hunk 7]
```diff
@@ -33,21 +34,32 @@ def __init__(
         timeout: int = 30,
         verify_ssl: bool = True,
         max_concurrent_requests: int = 10,
+        custom_headers: Optional[Dict[str, str]] = None,
     ):
         self.base_url = base_url.rstrip("/")
         self.auth_type = auth_type
         self.auth_config = auth_config
         self.pagination_config = pagination_config
         self.timeout = timeout
```
Comment (emekanwaoma) [inline integrations/custom/http_server/client.py]: the client takes takes timeout, but OceanAsyncClient is created with timeout=ocean.config.client_timeout. The provided timeout is unused.

File: integrations/custom/main.py [hunk 8]
```diff
@@ -51,44 +53,51 @@ async def resync_resources(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
...
-
-                    if processed_batch:
-                        logger.info(
-                            f"Extracted {len(processed_batch)} items using data_path: {data_path}"
-                        )
-                        yield processed_batch
-                else:
+                # If data_path is default ('.') and batch[0] is not a list, yield as-is
+                if data_path == "." and batch and not isinstance(batch[0], list):
+                    logger.error(
```
Comment (emekanwaoma) [inline integrations/custom/main.py:58]: should this be a warning and not an error logging?

File: integrations/jira-server/poetry.lock [hunk 9]
```diff
@@ -1234,13 +1234,13 @@ testing = ["coverage", "pytest", "pytest-benchmark"]
 
 [[package]]
 name = "port-ocean"
-version = "0.29.6"
+version = "0.29.7"
```
Comment (emekanwaoma) [inline integrations/jira-server/poetry.lock:1237]: I think this this integration was touched in error. Best to remove from changes

File: integrations/custom/.port/resources/blueprints.json
Comment (omby8888) [reply integrations/custom/.port/resources/blueprints.json:1]: let's delete this file entirely as `[]` is the default

File: integrations/custom/main.py (hunk 8 above)
Comment (Matarp13) [reply integrations/custom/main.py:58]: Error, as this will cause the integration to not work
